src/solvers/refinement/string_constraint_generator_format.cpp

```diff
@@ -0,0 +1,424 @@
+  case format_specifiert::DECIMAL_FLOAT:
+  case format_specifiert::HEXADECIMAL_FLOAT:
+    return add_axioms_for_string_of_float(
+      get_component_in_struct(arg, ID_float), ref_type);
+  case format_specifiert::CHARACTER:
+    return add_axioms_from_char(
+      get_component_in_struct(arg, ID_char), ref_type);
+  case format_specifiert::BOOLEAN:
+    return add_axioms_from_bool(
+      get_component_in_struct(arg, ID_boolean), ref_type);
+  case format_specifiert::STRING:
+    return get_string_expr(get_component_in_struct(arg, "string_expr"));
+  case format_specifiert::HASHCODE:
+    // TODO: no implementation for hashcode
+    return add_axioms_from_int(
+      get_component_in_struct(arg, "hashcode"), MAX_INTEGER_LENGTH, ref_type);
+  case format_specifiert::LINE_SEPARATOR:
+    // TODO: the constant should depend on the system: System.lineSeparator()
+    return add_axioms_for_constant("\n", ref_type);
+  case format_specifiert::PERCENT_SIGN:
+    return add_axioms_for_constant("%", ref_type);
+  case format_specifiert::SCIENTIFIC_UPPER:
+  case format_specifiert::GENERAL_UPPER:
+  case format_specifiert::HEXADECIMAL_FLOAT_UPPER:
+  case format_specifiert::CHARACTER_UPPER:
+  case format_specifiert::DATE_TIME_UPPER:
+  case format_specifiert::BOOLEAN_UPPER:
+  case format_specifiert::STRING_UPPER:
+  case format_specifiert::HASHCODE_UPPER:
+  {
+    string_constraint_generatort::format_specifiert fs_lower=fs;
+    fs_lower.conversion=tolower(fs.conversion);
+    string_exprt lower_case=add_axioms_for_format_specifier(
+      fs_lower, arg, ref_type);
+    return add_axioms_for_to_upper_case(lower_case);
+  }
+  case format_specifiert::DATE_TIME:
+  default:
+    // TODO: printDateTime is not implemented
+    throw std::string("invalid format specifier:").append(1, fs.conversion);
+  }
+}
+
+/// Parse `s` and add axioms ensuring the output corresponds to the output of
+/// String.format.
+/// \param s: a format string
+/// \param args: a vector of arguments
+/// \param ref_type: a type  for refined string type
+/// \return String expression representing the output of String.format.
+string_exprt string_constraint_generatort::add_axioms_for_format(
+  const std::string &s,
+  const std::vector<exprt> &args,
+  const refined_string_typet &ref_type)
+{
+  const std::vector<format_elementt> format_strings=parse_format_string(s);
+  std::vector<string_exprt> intermediary_strings;
+  std::size_t arg_count=0;
+
+  for(const format_elementt &fe : format_strings)
+    if(fe.is_format_specifier())
+    {
+      // TODO we shouldn't increment arg_count for constants
+      const format_specifiert &fs=fe.get_format_specifier();
+      struct_exprt arg;
+      if(fs.conversion!=format_specifiert::PERCENT_SIGN &&
+         fs.conversion!=format_specifiert::LINE_SEPARATOR)
+        arg=to_struct_expr(args[arg_count++]);
+      intermediary_strings.push_back(
+        add_axioms_for_format_specifier(fs, arg, ref_type));
```

Comment:
`add_axioms_for_format_specifier` does not use the argument if `fs` is one of the two above, so we are fine